### ISSUE TITLE
\showstream and \ShowHook

### ISSUE THREAD

**Issue by Rimole** [latex3/latex2e#1062]
## Brief outline of the enhancement request
The output of \showstream should either go to the file, or the log, or both, but not partially to the file and partially to the log (as it currently does).

"\ShowHook is a mixture of (essentially) \wlog and \show to output  some latex information then finally give the \show interaction, 
\showstream [...] only affects the parts using \show, so the message preamble using \write still goes to the log. Latex could (I suppose) check every time if \showstream is an open stream and if so write to there, or perhaps always write to \showstream and rely on tex's behaviour of writing to the log for unopened stream numbers." (David Carlisle)

## Minimal example showing the current behaviour

```latex
\RequirePackage{latexbug}
\documentclass{minimal}
\newwrite\teststream
\immediate\openout\teststream="teststream"
\showstream=\teststream
\def\example{\message{This is a test.}}
\show\example
\AddToHook{begindocument}[foobar]{\message{Hi!}}
\ShowHook{begindocument}
\begin{document}
\end{document}
```

## Log file and testsream.tex file  
[Test.log](https://github.com/latex3/latex2e/files/11472112/Test.log)

[teststream.tex.txt](https://github.com/latex3/latex2e/files/11472114/teststream.tex.txt)

(.tex files not accepted for upload, I appended a .txt)

**Comment by FrankMittelbach** [comment]
@PhelypeOleinik you intend to fix that for the fall release?

**Comment by PhelypeOleinik** [comment]
I do. I started this for the now-released release, but I got caught with cmd args halfway through.

**Comment by muzimuzhi** [comment]
Currently `\ShowHook`/`\hook_show:n` uses a mixture of `\iow_log:x` and `\tl_show:n`.

`\iow_log:x` uses the constant output stream `\c_log_iow` while `\tl_show:n` uses `\tex_showtokens:D`/`\showtokens` which respects `\showstream`.

I've opened a feature request in latex3 repository, see latex3/latex3#1532.

**Comment by josephwright** [comment]
The work is now done for this in `expl3`, so we can look to implement - the only question is whether we need to worry about the case of older `expl3` with an updated kernel. My feeling is not - this is a debugging feature so we can be a bit more 'aggressive' - but I welcome input.

**Comment by FrankMittelbach** [comment]
I don't think we should worry about this, especially because it is just offering a debugging possibility (but even if not: an up-to-date kernel should be run with an up-to-date L3 layer).

**Comment by josephwright** [comment]
Mumble, implementation issue in the `expl3` code to fix first
